api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,207 @@
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"go.opencensus.io/trace"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+// CloudEvent is the official JSON representation of a CloudEvent: https://github.com/cloudevents/spec/blob/master/serialization.md
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+type cloudEventIn struct {
+	CloudEvent
+
+	// Deadline string `json:"deadline"`
+	// Protocol CallRequestHTTP `json:"protocol"`
+}
+
+// cloudEventOut the expected response from the function container
+type cloudEventOut struct {
+	CloudEvent
+
+	// Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
+}
+
+// CloudEventProtocol converts stdin/stdout streams from HTTP into JSON format.
+type CloudEventProtocol struct {
+	// These are the container input streams, not the input from the request or the output for the response
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *CloudEventProtocol) IsStreamable() bool {
+	return true
+}
+
+func (h *CloudEventProtocol) writeJSONToContainer(ci CallInfo) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	_, err := io.Copy(buf, ci.Input())
+	if err != nil {
+		return err
+	}
+
+	var in cloudEventIn
+	// fmt.Printf("is cloudevent? %v\n", ci.IsCloudEvent())
+	if ci.IsCloudEvent() {
+		// then it's already in the right format, let's parse it, then modify
+		err = json.Unmarshal(buf.Bytes(), &in)
+		if err != nil {
+			return fmt.Errorf("Invalid CloudEvent input. %v", err)
+		}
```

Comment:
if they send in a cloud event we aren't setting it to the call id, they could leave it blank or use their own (other relevant fields like content-type, event-type, cloud-events-version, source). i think this is the intent most likely of the spec (to let the caller set these) and since this has mixed responsibilities at the moment of generating cloud events or taking them, maybe this is ok, just pointing this out if we want to fix this.

Reply:
Ya, the caller should have at least set the required fields. We'll probably have to store a separate call_id and event_id field in our db to correlate, but we can get to that in fn2.0.